### PR TITLE
Update mapping support in qchem

### PR DESCRIPTION
**Pull request by ddhawan11**
**Context:**
added different mappings to molecular_hamiltonian function

**Description of the Change:**
molecular Hamiltonian can be mapped using different mapping schemes.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 1]
```diff
@@ -153,11 +153,12 @@ def test_building_hamiltonian_molecule_class(
 
 
 @pytest.mark.parametrize(
-    ("symbols", "geometry", "h_ref_data"),
+    ("symbols", "geometry", "mapping", "h_ref_data"),
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:156]: We might need to add "method" here to make sure we check both "dhf" and "pyscf" backends for different mappings.

**Review by soranjh — COMMENTED** [review]
Thanks @ddhawan11, looks good overall. We need to make sure that the pyscf backend is also checked with different mappings.

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 2]
```diff
@@ -202,9 +203,106 @@ def test_building_hamiltonian_molecule_class(
                 ],
             ),
         ),
+        (
+            ["H", "H"],
+            np.array([0.0, 0.0, 0.0, 0.0, 0.0, 1.0]),
+            "parity",
+            # computed with OpenFermion; data reordered
```
[the remark this answers is not in the document]
Comment (ddhawan11) [reply tests/qchem/of_tests/test_molecular_hamiltonian.py:209]: Reference Hamiltonian is computed with Openfermion, this line stayed as a copy/paste line though.

File: pennylane/qchem/openfermion_obs.py [hunk 3]
```diff
@@ -1123,18 +1125,36 @@ def _molecular_hamiltonian(
...
         hf = qml.qchem.fermionic_observable(core_constant, one_mo, two_mo)
 
         if active_new_opmath():
-            h_pl = qml.jordan_wigner(hf, wire_map=wires_map, tol=1.0e-10).simplify()
-
+            if mapping.strip().lower() == "jordan_wigner":
+                h_pl = qml.jordan_wigner(hf, wire_map=wires_map, tol=1.0e-10).simplify()
+            elif mapping.strip().lower() == "parity":
+                qubits = len(hf.wires)
+                h_pl = qml.jordan_wigner(hf, qubits, wire_map=wires_map, tol=1.0e-10).simplify()
```
[the remark this answers is not in the document]
Comment (ddhawan11) [reply pennylane/qchem/openfermion_obs.py]: Thanks for noticing, fixed now, added more tests as well.

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 4]
```diff
@@ -202,9 +203,106 @@ def test_building_hamiltonian_molecule_class(
...
+            # h_mol = molecule.get_molecular_hamiltonian()
+            # h_f = openfermion.transforms.get_fermion_operator(h_mol)
+            # h_q = openfermion.transforms.jordan_wigner(h_f)
```
[the remark this answers is not in the document]
Comment (ddhawan11) [reply tests/qchem/of_tests/test_molecular_hamiltonian.py]: This is corrected now.

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 5]
```diff
@@ -202,9 +203,106 @@ def test_building_hamiltonian_molecule_class(
...
+            ),
+        ),
+        (
+            ["H", "H"],
+            np.array([0.0, 0.0, 0.0, 0.0, 0.0, 1.0]),
+            "bravyi_kitaev",
+            # computed with OpenFermion; data reordered
+            # h_mol = molecule.get_molecular_hamiltonian()
+            # h_f = openfermion.transforms.get_fermion_operator(h_mol)
+            # h_q = openfermion.transforms.jordan_wigner(h_f)
```
[the remark this answers is not in the document]
Comment (ddhawan11) [reply tests/qchem/of_tests/test_molecular_hamiltonian.py]: This is corrected now.

File: molecule_pyscf_sto-3g.hdf5
Comment (trbromley) [inline molecule_pyscf_sto-3g.hdf5:1]: Is this file expected?

Comment (ddhawan11) [reply molecule_pyscf_sto-3g.hdf5:1]: Thanks for noticing. This was getting created with some tests, it is fixed now.

File: pennylane/qchem/openfermion_obs.py [hunk 6]
```diff
@@ -1123,18 +1122,36 @@ def _molecular_hamiltonian(
...
+            elif mapping.strip().lower() == "parity":
+                qubits = len(hf.wires)
+                h_pl = qml.parity_transform(
+                    hf, qubits, ps=True, wire_map=wires_map, tol=1.0e-10
+                ).hamiltonian()
+            elif mapping.strip().lower() == "bravyi_kitaev":
+                qubits = len(hf.wires)
+                h_pl = qml.bravyi_kitaev(
+                    hf, qubits, ps=True, wire_map=wires_map, tol=1.0e-10
+                ).hamiltonian()
```
Comment (austingmhuang) [inline pennylane/qchem/openfermion_obs.py:1153]: This long if else block contains a lot of repeated work and could be simplified to be made more readable. I think it's best to define things like mapping.strip.lower() or the # of qubits once and use it throughout the if-else block.

**Review by austingmhuang — COMMENTED** [review]
Good work @ddhawan11. Couple of changes but otherwise I think it's good.

File: tests/qchem/of_tests/test_decompose.py [hunk 7]
```diff
@@ -539,17 +539,3 @@ def test_transformation(name, core, active, mapping, coeffs_ref, pauli_strings_r
 
     assert np.allclose(coeffs, coeffs_ref, **tol)
     assert pauli_strings == pauli_strings_ref
-
-
-@pytest.mark.usefixtures("skip_if_no_openfermion_support")
```
[the remark this answers is not in the document]
Comment (ddhawan11) [reply tests/qchem/of_tests/test_decompose.py:544]: This tests if a particular mapping is available for openfermion method, we test that for all methods in molecular_hamiltonian function now.

File: pennylane/qchem/openfermion_obs.py (hunk 6 above)
Comment (ddhawan11) [reply pennylane/qchem/openfermion_obs.py]: I have cleaned it up some, but I don't want to use ternary operators as that can make code hard to read with nested if statements.

**Review by soranjh — APPROVED** [review]
Thanks @ddhawan11, looks good. Please update the changelog before merging.

**Review by austingmhuang — APPROVED** [review]
Approved. Great work @ddhawan11 !

**Pull request merged**